houdini: 18.5.596 -> 19.5.569

permahorse opened the pull request:
###### Description of changes

* added missing libs to buildHFSUserEnv's targetPkgs
* exposed more useful executables provided by houdini
* binding /run/opengl-driver/etc/OpenCL/vendors -> /etc/OpenCL/vendors  to use opencl with houdini's own version of libOpenCL.so

~~does not work with new buildFHSEnv from #225748 , so using buildFHSEnvChroot instead~~

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Atemu commented on pkgs/applications/misc/houdini/default.nix:
```diff
@@ -1,22 +1,45 @@
-buildFHSEnv rec {
+{ lib, stdenv, writeScript, callPackage, buildFHSEnvChroot, bubblewrap, unwrapped ? callPackage ./runtime.nix {} }:
+let
+  make_empty_ocl_in_etc = stdenv.mkDerivation{  # a cheat to have OpenCL/vendors dir to bind at runtime
```
`pkgs.runCommand`

pedohorse replied:
fixed

Atemu commented on pkgs/applications/misc/houdini/default.nix:
```diff
@@ -1,22 +1,45 @@
-    libxkbcommon libudev0-shim tbb
+    libxkbcommon libudev0-shim tbb xwayland qt5.qtwayland nettools
+    bintools  # needed for ld and other tools, so ctypes can find/load sos from python
+    ocl-icd  # needed for opencl
+    numactl ncurses5  # needed by hfs ocl backend
   ] ++ (with xorg; [
-    libICE libSM libXmu libXi libXext libX11 libXrender libXcursor libXfixes
+    libICE libSM libXmu libXi libXt libXext libX11 libXrender libXcursor libXfixes
     libXrender libXcomposite libXdamage libXtst libxcb libXScrnSaver
-  ]);
+    libXrandr
+    libxcb xcbutil xcbutilimage xcbutilrenderutil xcbutilcursor xcbutilkeysyms xcbutilwm
+  ]) ++ [
+    make_empty_ocl_in_etc
+  ];
```
These are becoming unwieldy, perhaps now would be a good time to split this across multiple lines.

pedohorse replied:
what do you suggest? having one per line seem excessive and less readable

Atemu replied:
One per line is what people prefer these days.

IMO it creates _much_ more readable diffs and I don't think it hurts readability. If I see a long vertical stack of package names, I immediately recognise it as a list of buildInputs and can trivially mentally skip it if I don't care about buildInputs right now.

Atemu replied:
Best try it for yourself. If you're going to do that, please do it in a separate commit so it's easy to tell which deps you added.

Atemu commented on pkgs/applications/misc/houdini/default.nix:
```diff
@@ -32,6 +55,10 @@ buildFHSEnv rec {
 
   runScript = writeScript "${name}-wrapper" ''
-    exec $@
+    if [ -d /run/opengl-driver/etc/OpenCL/vendors ]; then
+      exec ${bubblewrap}/bin/bwrap --dev-bind / / --bind /run/opengl-driver/etc/OpenCL/vendors $(realpath /etc/OpenCL)/vendors $@
+    else
+      exec $@
+    fi
```
I don't a nested bubblewrap is a good solution here.

Also, if bwrap works here, why use buildFHSEnvChroot?

pedohorse replied:
bwrap here is only needed to bind OpenCL/vendors to the ones in running system. 

I just did not find another way to do it, since it cannot be done during build. Do you have suggestions?

So with `buildFHSEnvChroot` and this one bwrap there are no nested bwraps, right?

as for why `buildFHSEnv` does not work - i seem to have figured it out. the reason is that it has `--die-with-parent` option passed to bwrap, and houdini loves to spawn processes in bg. Even by default it forks and exits shortly after start, which leads to bwrap parent process exiting and sending SIGKILL to houdini. Also it spawns hserver process that is meant to be accessed by multiple houdini instances, but having `--unshare-pid` option makes houdini fail to find it if starting several instances in parallel, which leads to licensing errors.... Long story short - i think it's doable to switch to `buildFHSEnv`, but i'd stick with the Chroot one for now, as it behaves in a more expected way.

pedohorse replied:
nice! I've only looked into manual and haven't seen these extra arguments (still pretty new to nix in general)

switched to `buildFHSEnv` now, everything seem to work fine

also doing a `--ro-bind-try /etc/OpenCL/vendors /etc/OpenCL/vendors` to address non-NixOS case.  
Assuming both `/etc/OpenCL/vendors` and  `/run/opengl-driver/etc/OpenCL/vendors/` won't ever be present on a system at the same time

Atemu commented on pkgs/applications/misc/houdini/default.nix:
```diff
@@ -1,22 +1,36 @@
-{ lib, stdenv, writeScript, callPackage, buildFHSEnv, unwrapped ? callPackage ./runtime.nix {} }:
-
+{ lib, stdenv, runCommand, writeScript, callPackage, buildFHSEnv, bubblewrap, unwrapped ? callPackage ./runtime.nix {} }:
```
```suggestion
{ lib, stdenv, writeScript, callPackage, buildFHSEnv, unwrapped ? callPackage ./runtime.nix {} }:
```

Atemu reviewed (approved):
Diff LGTM now.

I can't test that this works but I assume it was broken and I trust you tested it, so I'll merge in a few days if nobody objects.

Atemu commented:
While update commits can certainly include all the changes required for the update compared to the previous version, in this case the previous version was already broken (I assume?). I think I'd prefer you put the things you did into separate commits. Fix up the drv first, then upgrade to v19 + things that are different for v19 specifically.

Commit:
houdini: expanded lists

one entry per line for targetPkgs, executables - for simpler diffs

pedohorse commented:
> While update commits can certainly include all the changes required for the update compared to the previous version, in this case the previous version was already broken (I assume?). I think I'd prefer you put the things you did into separate commits. Fix up the drv first, then upgrade to v19 + things that are different for v19 specifically.

split the update into 2 commits: one fixes FHS env, second updates version.
i have not tried to build the 18.5 original package, but i bet it was not fully functional, as for example opencl parts and karma parts did not change that much in houdini from 18.5->19.5.  
therefore i've just packed all changes to FHS in one commit (not splitting into something that fixes particularly version 18.5 and something new required only in 19.5)

Atemu reviewed (changes requested):
Please split the second commit into four and adhere to the commit message guidelines.

SuperSandro2000 commented on pkgs/applications/misc/houdini/runtime.nix:
```diff
@@ -22,10 +22,10 @@ stdenv.mkDerivation rec {
                       $out
     echo "licensingMode = localValidator" >> $out/houdini/Licensing.opt
-  '';
+  '';  # last localValidator line does not seem to do anything any more. not sure, official docs do not say anything about it
```
Please write the comment on that line

pedohorse replied:
moved the comment

SuperSandro2000 commented on pkgs/applications/misc/houdini/default.nix:
```diff
@@ -1,22 +1,75 @@
 buildFHSEnv rec {
   name = "houdini-${unwrapped.version}";
 
+  dieWithParent = false;  # houdini spawns hserver (and other license tools) that is supposed to live beyond the lifespan of houdini process
+
+  unsharePid = false;  # houdini needs to communicate with hserver process that it seem to be checking to be present in running processes
```
```suggestion
  # houdini spawns hserver (and other license tools) that is supposed to live beyond the lifespan of houdini process
  dieWithParent = false;

  # houdini needs to communicate with hserver process that it seem to be checking to be present in running processes
  unsharePid = false;
```

pedohorse replied:
fixed

Commit:
houdini: added missing libs to targetPkgs

Commit:
houdini: disabled dieWithParent, unsharePid

Due to the nature of the software, it forks and expects children to
outlive the parent, therefore dieWithParent needs to be disabled
Also licensing mechanism seem to be checking processes for it's
licensing helper process running. that process must be same for multiple
instances of houdini itself, therefore Pid namespace has to be the same
for all of them.

Commit:
houdini: exposed more executables

houdini provides additional tools that might be useful

Commit:
houdini: added /etc/OpenCL/vendors binding

additionally trying to bind
/run/opengl-driver/etc/OpenCL/vendors -> /etc/OpenCL/vendors
/etc/OpenCL/vendors -> /etc/OpenCL/vendors (for non NixOS)
in wrapper to use opencl with houdini's own version of libOpenCL.so

Commit:
houdini: 18.5.596 -> 19.5.569

pedohorse commented:
> Please split the second commit into four and adhere to the commit message guidelines.

commit split. messages adjusted

Atemu commented:
Nit: Commit messages are usually written in simple past.

Atemu commented:
Um, sorry, simple present. I was tired.
